libvmi/os/linux/elfparse.c

```diff
@@ -0,0 +1,422 @@
+        return VMI_FAILURE;
+    }
+
+    return VMI_SUCCESS;
+}
+
+status_t
+elfparse_get_image(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    size_t len,
+    const uint8_t * const image)
+{
+    if (VMI_FAILURE == vmi_read(vmi, ctx, len, (void *)image, NULL)) {
+        dbprint(VMI_DEBUG_ELFPARSE, "--ELFPARSE: failed to read ELF header\n");
+        return VMI_FAILURE;
+    }
+
+    if (VMI_SUCCESS != elfparse_validate_elf_image(image, len)) {
+        dbprint(VMI_DEBUG_ELFPARSE, "--ELFPARSE: failed to validate ELF header\n");
+        return VMI_FAILURE;
+    }
+
+    return VMI_SUCCESS;
+}
+
+static inline uint32_t elf_hash(const uint8_t* name)
+{
+    uint32_t h = 0, g;
+
+    for (; *name; name++) {
+        h = (h << 4) + *name;
+        if ((g = h & 0xf0000000))
+            h ^= g >> 24;
+        h &= ~g;
+    }
+
+    return h;
+}
+
+static status_t elf_hash_lookup(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    addr_t str_tab,
+    addr_t sym_tab,
+    addr_t hash_tab,
+    const char *name,
+    addr_t *rva)
+{
+    access_context_t _ctx = *ctx;
+    uint32_t hash = elf_hash((uint8_t*) name);
+    uint32_t table[2];
+
+    _ctx.addr = hash_tab;
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(table), table, NULL))
+        return VMI_FAILURE;
+
+    uint32_t current;
+    _ctx.addr += sizeof(table) + sizeof(uint32_t) * (hash % table[0]);
+    if (VMI_FAILURE == vmi_read_32(vmi, &_ctx, &current))
+        return VMI_FAILURE;
+
+    while (current) {
+        struct elf64_sym sym;
+        _ctx.addr = sym_tab + sizeof(struct elf64_sym) * current;
+        if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(struct elf64_sym), &sym, NULL))
+            break;
+
+        _ctx.addr = str_tab + sym.st_name;
+        char *n = vmi_read_str(vmi, &_ctx);
+        if (n && !strcmp(name, n)) {
+            free(n);
+            *rva = sym.st_value;
+            return VMI_SUCCESS;
+        } else if (n)
+            free(n);
+
+        _ctx.addr = hash_tab + sizeof(table) + sizeof(uint32_t) * (table[0] + current);
+        if (VMI_FAILURE == vmi_read_32(vmi, &_ctx, &current))
+            break;
+    }
+
+    return VMI_FAILURE;
+}
+
+static uint32_t elf_hash_max_symbols(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    addr_t hash_tab,
+    uint32_t *max_symbols)
+{
+    access_context_t _ctx = *ctx;
+    uint32_t table[2];
+
+    _ctx.addr = hash_tab;
+    if (VMI_FAILURE == vmi_read(vmi, &_ctx, sizeof(table), table, NULL))
+        return VMI_FAILURE;
+
+    *max_symbols = table[1];
+    return VMI_SUCCESS;
+}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstack-protector"
+
+void
+elfparse_assign_tables(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    const uint8_t * const image,
+    addr_t *hash,
+    addr_t *gnu_hash,
+    addr_t *str,
+    addr_t *sym)
+{
+    struct elf64_ehdr *elf = (struct elf64_ehdr *) image;
+    struct elf64_phdr ph[elf->e_phnum];
```

Comment:
I did a bit of testing on this and the reason why `-Wstack-protector` trips on this function is this above. You are allocating an array on the stack that's variable size. That's not a good thing. You aren't sanity checking that `elf->e_phnum` is within a reasonable range either. Considering the header may be entirely malicious or malformed, that's a concern. You should switch this to a `g_slice_alloc0` after sanity checking and making sure it gets `g_slice_free`d before the function returns.

Reply:
I had already changed that in 2cf503a but reverted it back since the other warnings were still there. I'm not too happy with allocating it on the heap. That seems quite inefficient to me. Would you be happy with the sanity check and a fixed size allocation on the stack?